polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp

```diff
@@ -1716,6 +1717,270 @@ class RaiseSetCaptured : public OpHostRaisePattern<LLVM::VarAnnotation> {
+        continue;
+      if (auto v = TypeSwitch<Operation *, Value>(user)
+                       .Case<LLVM::StoreOp>([&](auto st) {
+                         return st.getAddr() == ptr ? st.getValue() : Value();
+                       })
+                       .Default(Value())) {
+        if (value)
+          // Assignment is not unique.
+          return {};
+
+        value = v;
+      } else {
+        // Marker intrinsics are unproblematic.
+        if (isa<LLVM::LifetimeStartOp, LLVM::LifetimeEndOp>(user))
+          continue;
+
+        // Unexpected user; conservatively assume that assignment is not unique.
+        return {};
+      }
+    }
+
+    return value;
+  }
+
+  /// Scan \p block for `llvm.intr.var.annotation`s to try to extract the number
+  /// of kernel parameters and a pointer to the first descriptor.
+  static FailureOr<std::tuple<unsigned, Value>>
+  getKernelParameterInfo(Block *block) {
+    auto annotations = block->getOps<LLVM::VarAnnotation>();
+    if (std::distance(annotations.begin(), annotations.end()) != 2)
+      return failure();
+
+    LLVM::VarAnnotation numParamsAnn = *annotations.begin();
+    LLVM::VarAnnotation paramDescAnn = *++annotations.begin();
+    FailureOr<StringRef> numParamsAnnStr = getAnnotation(numParamsAnn);
+    FailureOr<StringRef> paramDescAnnStr = getAnnotation(paramDescAnn);
+    if (failed(numParamsAnnStr) || failed(paramDescAnnStr) ||
+        *numParamsAnnStr != "kernel_num_params" ||
+        *paramDescAnnStr != "kernel_param_desc")
+      return failure();
+
+    Value numParams = getAnnotatedValue(numParamsAnn);
+    Value paramDesc = getAnnotatedValue(paramDescAnn);
+
+    APInt numParamsConst;
+    if (!matchPattern(numParams, m_ConstantInt(&numParamsConst)))
+      return failure();
+
+    return std::tuple<unsigned, Value>{numParamsConst.getZExtValue(),
+                                       paramDesc};
+  }
+
+  /// Try to parse \p numParams kernel parameter descriptors starting from the
+  /// \p paramDesc pointer into the `kernel_signatures` constant from the
+  /// integration header. If successful, return the descriptors' kinds.
+  static FailureOr<SmallVector<kernel_param_kind_t>>
+  getParameterKinds(unsigned numParams, Value paramDesc) {
+    FlatSymbolRefAttr kernelSignaturesRef;
+    unsigned offset = 0;
+    auto addressOf = paramDesc.getDefiningOp<LLVM::AddressOfOp>();
+    if (addressOf)
+      kernelSignaturesRef = addressOf.getGlobalNameAttr();
+    else {
+      auto gep = paramDesc.getDefiningOp<LLVM::GEPOp>();
+      if (!gep || !gep.getDynamicIndices().empty())
+        return failure();
+
+      auto indices = gep.getRawConstantIndices();
+      if (indices.size() != 2 || indices[0] != 0)
+        return failure();
+      offset = indices[1];
+
+      addressOf = gep.getBase().getDefiningOp<LLVM::AddressOfOp>();
+      if (!addressOf)
+        return failure();
+      kernelSignaturesRef = addressOf.getGlobalNameAttr();
+    }
```

Comment:
```suggestion
    if (auto addressOf = paramDesc.getDefiningOp<LLVM::AddressOfOp>()) {
      kernelSignaturesRef = addressOf.getGlobalNameAttr();
    } else {
      auto gep = paramDesc.getDefiningOp<LLVM::GEPOp>();
      if (!gep || !gep.getDynamicIndices().empty())
        return failure();

      auto indices = gep.getRawConstantIndices();
      if (indices.size() != 2 || indices[0] != 0)
        return failure();
      offset = indices[1];

      addressOf = gep.getBase().getDefiningOp<LLVM::AddressOfOp>();
      if (!addressOf)
        return failure();
      kernelSignaturesRef = addressOf.getGlobalNameAttr();
    }
```

I think this improves readability